pygfx/utils/show.py

```diff
@@ -16,9 +17,243 @@
     OrbitController,
     AmbientLight,
     DirectionalLight,
+    Light,
+    Camera,
+    Controller,
 )
 
 
+class RenderCallback:
+    """A callback used during the draw call
+
+    This class implements a Descriptor that manages the draw call callbacks of
+    Display. It's main purpose is to allow setting the default value of these
+    callbacks using a decorator-style syntax::
+
+        @gfx.Display.before_render
+        def animate():
+            ...
+
+        gfx.show()  # uses animate
+
+    Parameters
+    ----------
+    default : Callable
+        The default callback to use.
+
+    Notes
+    -----
+    Using the decorator feature above sets the class-level default for the
+    callback and affects all instances of Display that use the default value.
+
+    """
+
+    def __init__(self, default=None) -> None:
+        self.default_callback = default
+
+    def __set_name__(self, owner, name):
+        self.private_name = "_" + name
+
+    def __get__(self, instance, type=None):
+        # allow decorator-style setting of default
+        # callback
+        if instance is None:
+            return self
+
+        value = getattr(instance, self.private_name, None)
+        if value is not None:
+            return value
+
+        # persist callback to prevent it from changing
```

Comment:
Can you explain how this prevents it? I'm not super clear on the control flow here

Reply:
Right. Under the hood, the descriptor uses a `_<name>` attribute on the instance to manage the actual callback, i.e., if we do

```python
disp = Display()

disp.before_render = animate
```

then the descriptor creates `disp._before_render` to store the callback on `disp`. If that value doesn't exist or is `None` we will fall back to `gfx.Display.before_render.default_callback`. This is what the `if value is None: ...` does.

This works for most cases, except when we have an already running `disp` (e.g. inside an earlier notebook cell) that uses the default callback. If, in this case, we use the `@gfx.Display.before_render` decorator while setting up the next display object we may modify the running one. Since the decorator sets the class-level default (which the previous `disp` is using) we risk the rendering loop of the running renderer changing. To prevent this interaction I take the value `Display.before_render.default_callback` during the first run and persist it on the instance by setting `disp._before_render = Display.before_render.default_callback`. Now future changes of the class-level default can no longer impact the instance-level `disp` once the display starts running/showing.

Writing this out makes two enhancements come to mind: (1) We probably want to not use `default_callback` if the user has explicitly set `disp.before_render = None`. (2) We could add a `Display.finalize()` function that does all the setup that can be done without making a canvas pop up. I'm thinking in particular of setting all the callbacks and making them independent of the class-level defaults. The second one is not needed right now, but is quite easy to incorporate.

Reply:
👍 thanks for explaining

Reply:
No problem. Should I add a clarifying comment in the source for future reference somewhere? (I also did a refactor of this section just now, so it might be a little more intuitive now)